packages/modules/product/src/repositories/product-category.ts

```diff
@@ -86,14 +88,31 @@ export class ProductCategoryRepository extends DALUtils.MikroOrmBaseTreeReposito
       return productCategories
     }
 
-    return await this.buildProductCategoriesWithTree(
+    const categoriesTree = await this.buildProductCategoriesWithTree(
       {
         descendants: transformOptions.includeDescendantsTree,
         ancestors: transformOptions.includeAncestorsTree,
       },
       productCategories,
       findOptions_
     )
+
+    return this.sortCategoriesByRank(categoriesTree)
+  }
+
+  sortCategoriesByRank(categories: ProductCategory[]): ProductCategory[] {
+    const sortedCategories = categories.sort((a, b) => a.rank - b.rank)
+
+    for (const category of sortedCategories) {
+      if (category.category_children) {
+        // All data up to this point is manipulated as an array, but it is a Collection<ProductCategory> type under the hood, so we are casting to any here.
+        category.category_children = this.sortCategoriesByRank(
+          category.category_children as any
```

Comment:
Q: I belive you can turn it to an array by doing`[...category.category_children]` or `Array.from(category.category_children)` let me know what you think

Reply:
it is actually an array already (if you see how the data is created in `buildProductCategoriesWithTree`), it's just that the typings are not incorrect. But I'll try to clean it up as you suggested